lib/src/git.rs

```diff
@@ -1812,3 +1813,145 @@ pub fn parse_gitmodules(
         .collect();
     Ok(ret)
 }
+
+#[derive(Error, Debug)]
+pub enum ColocatedWorkspaceWarning {
+    #[error("Broken .git symlink, pointing to {}", target.display())]
+    BrokenGitSymlink { target: PathBuf },
+}
+
+impl ColocatedWorkspaceWarning {
+    pub fn hint(&self) -> Option<String> {
+        match self {
+            _ => None,
+        }
+    }
+}
+
+/// How the git repository was opened
+#[derive(Debug)]
+pub enum GitRepoColocationType {
+    /// Internal repo managed by JJ, or an external git repo
+    NotColocated {
+        /// Colocation is auto-detected. This indicates whether we noticed
+        /// a problem with the colocation setup, and ignored it, falling back
+        /// to NotColocated.
+        warning: Option<ColocatedWorkspaceWarning>,
+    },
+
+    /// Repo stored in a workspace root .git directory, such that git tooling
+    /// also works
+    Colocated,
+}
+
+impl GitRepoColocationType {
+    pub(crate) const fn internal() -> Self {
+        Self::NotColocated { warning: None }
+    }
+    pub fn is_colocated(&self) -> bool {
+        matches!(self, Self::Colocated)
+    }
+    pub fn warning(&self) -> Option<&ColocatedWorkspaceWarning> {
```

Comment:
nit: as I said in other thread, I think warning stuff can be moved to CLI. The `GitBackend` doesn't have to know that there was a `.git` directory that looked like a colocated repo but it wasn't.